### DISCUSSION TITLE
React router dom

### DISCUSSION THREAD

**Discussion by startthecode** [community/community#116446] [Programming Help] [answered]
`import {
  Routes,
  Route,
  useLocation,
  createBrowserRouter,
  createRoutesFromElements,
} from "react-router-dom";
import { AppLayout } from "../Components/layout/AppLayout";
import { lazy, Suspense } from "react";
let About = lazy(() => import("../pages/About.jsx"));
let Home = lazy(() => import("../pages/Home.jsx"));
let SignIn = lazy(() => import("../pages/SignIn.jsx"));
let SignUp = lazy(() => import("../pages/SignUp.jsx"));

export const AllRoutes = createBrowserRouter(
  createRoutesFromElements(
    <>
      <Route
        path=""
        element={
          <Suspense fallback="Loading">
            <AppLayout />
          </Suspense>
        }
      >
        <Route path="/" element={<Home />} />
        <Route path="/about" element={<About />} />
        <Route path="/contact" />
        <Route path="/work" />
      </Route>
      <Route
        path="/signin"
        element={
          <Suspense fallback="Loading">
            <SignIn />
          </Suspense>
        }
      />
      <Route
        path="/signup"
        element={
          <Suspense fallback="Loading">
            <SignUp />
          </Suspense>
        }
      />
    </>
  )
);
Hey everyone, I'm creating routes like this. Do you think this is a good approach?`

**Comment by ettaboyle** [comment]
Thanks for posting in the GitHub Community, @startthecode! 

We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the Feed category is solely related to conversations around the GitHub product Feed. This post makes more sense in our [Programming Help](https://github.com/orgs/community/discussions/categories/programming-help?discussions_q=is%3Aopen+sort%3Adate_created+category%3A%22Programming+Help%22) category. I've gone ahead and moved it for you. Good luck!

**Comment by Xhst** [comment]
Your approach to creating routes in React is a good start (you have left the element prop empty for your routes, I think it's done on purpose)

```
import {
  Routes,
  Route,
  useLocation,
  createBrowserRouter,
  createRoutesFromElements,
} from "react-router-dom";
import { AppLayout } from "../Components/layout/AppLayout";
import { lazy, Suspense } from "react";
let About = lazy(() => import("../pages/About.jsx"));
let Home = lazy(() => import("../pages/Home.jsx"));
let SignIn = lazy(() => import("../pages/SignIn.jsx"));
let SignUp = lazy(() => import("../pages/SignUp.jsx"));

export const AllRoutes = createBrowserRouter(
  createRoutesFromElements(
    <>
      <Route path="" element={<AppLayout />}>
        <Route path="/" element={<Home />} />
        <Route path="/about" element={<About />} />
        <Route path="/signin" element={<SignIn />} />
        <Route path="/signup" element={<SignUp />} />
      </Route>
    </>
  )
);
```
In this updated code, I’ve added the respective components to the element prop of each Route. Also, I’ve assumed that AppLayout is a layout component that you want to render on every route, so I’ve added it as the element for the base Route.

Remember to wrap your routes component with Suspense in the parent component to handle the loading state while the lazy-loaded components are being fetched. This approach allows you to leverage code-splitting, which can help reduce the initial load time of your app by only loading the code necessary for the initial route.

**Reply by startthecode** [reply]
Is there any issue with wrapping a lazy-loaded component like this?"
`   <Suspense fallback="Loading">
            <AppLayout />
          </Suspense>`

**Reply by Xhst** [reply]
This is actually the recommended way to handle loading states in a component that’s being lazy-loaded.

**Reply by startthecode** [reply]
`import {
  Routes,
  Route,
  useLocation,
  createBrowserRouter,
  createRoutesFromElements,
} from "react-router-dom";
import { AppLayout } from "../Components/layout/AppLayout";
import { lazy, Suspense } from "react";
let About = lazy(() => import("../pages/About.jsx"));
let Home = lazy(() => import("../pages/Home.jsx"));
let SignIn = lazy(() => import("../pages/SignIn.jsx"));
let SignUp = lazy(() => import("../pages/SignUp.jsx"));

export const AllRoutes = createBrowserRouter(
  createRoutesFromElements(
    <>
      <Route
        path=""
        element={
          <Suspense fallback="Loading">
            <AppLayout />
          </Suspense>
        }
      >
        <Route path="/" element={<Home />} />
        <Route path="/about" element={<About />} />
      </Route>
      <Route
        path="/signin"
        element={
          <Suspense fallback="Loading">
            <SignIn />
          </Suspense>
        }
      />
      <Route
        path="/signup"
        element={
          <Suspense fallback="Loading">
            <SignUp />
          </Suspense>
        }
      />
    </>
  )
);
`
this is fine?

**Reply by Xhst** [reply]
As i wrote earlier your approach is good, the only thing is that you left the element prop empty for your routes. The element prop is where you should specify the component that should be rendered when the route is matched.

**Reply by startthecode** [reply]
Okay, thank you for the reply. It means a lot to junior developers. The components were empty because I was in the process of creating them.

**Reply by startthecode** [reply]
Hi @Xhst, 
export const RbacRoute = ({ element: OriginalComponent, path, routeName }) => {
  let { userRole, permitedRoutes } = useContext(UserContext());
  return ()=><Route path={path} element={<OriginalComponent />} />;
};
Uncaught Error: \[RbacRoute\] is not a \<Route\> component. All component children of \<Routes\> must be a \<Route\> or \<React.Fragment\>
Sure, here's a refined version:

I want to return Routes with some modifications for role-based routes, but this component throws an error as mentioned above.

**Comment by developer-sumit** [comment, marked as answer]
```jsx
import React, { lazy, Suspense } from "react";
import { BrowserRouter as Router, Route, Routes } from "react-router-dom";
import { AppLayout } from "../Components/layout/AppLayout";

// Lazy-loaded components
const About = lazy(() => import("../pages/About.jsx"));
const Home = lazy(() => import("../pages/Home.jsx"));
const SignIn = lazy(() => import("../pages/SignIn.jsx"));
const SignUp = lazy(() => import("../pages/SignUp.jsx"));

const AllRoutes = () => {
  return (
    <Router>
      <AppLayout>
        <Suspense fallback={<div>Loading...</div>}>
          <Routes>
            <Route path="/" element={<Home />} />
            <Route path="/about" element={<About />} />
            <Route path="/signin" element={<SignIn />} />
            <Route path="/signup" element={<SignUp />} />
          </Routes>
        </Suspense>
      </AppLayout>
    </Router>
  );
};

export default AllRoutes;
```

Changes made:

1. Import React components directly from the 'react' package rather than using destructuring assignment from the 'react-router-dom' package.
2. Import `BrowserRouter` as `Router` for clarity.
3. Move the lazy loading and Suspense usage outside the createRoutesFromElements function for a cleaner and more readable structure.
4. Use arrow function syntax for the AllRoutes component definition.
5. Provide a fallback component for Suspense to display while the lazy-loaded components are loading. In this case, it's a simple "Loading..." message, but you can customize it according to your needs.

This rewritten version should have the same functionality as your original code but with some minor improvements for readability and best practices. Make sure to mark this as answered if you like this.

**Reply by startthecode** [reply]
Hi @developer-sumit ,
export const RbacRoute = ({ element: OriginalComponent, path, routeName }) => {
let { userRole, permitedRoutes } = useContext(UserContext());
return ()=><Route path={path} element={} />;
};
Uncaught Error: [RbacRoute] is not a <Route> component. All component children of <Routes> must be a <Route> or <React.Fragment>
Sure, here's a refined version:

I want to return Routes with some modifications for role-based routes, but this component throws an error as mentioned above.
